From 119549ba831afc10d27da2db9880babe01085686 Mon Sep 17 00:00:00 2001
From: Daniel Cederman <cederman@gaisler.com>
Date: Thu, 12 Jan 2017 16:21:31 +0100
Subject: [PATCH 16/50] rcc: Add RCC target

Introduces the -qbsp=bspname flag as used in BCC2. Does the same work as
-qrtems but also adds -B /path/to/bspname/lib.
---
 gcc/config.gcc           |  7 +++++++
 gcc/config/sparc/rcc.h   | 37 +++++++++++++++++++++++++++++++++++++
 gcc/config/sparc/t-rtems | 13 ++++++++-----
 3 files changed, 52 insertions(+), 5 deletions(-)
 create mode 100644 gcc/config/sparc/rcc.h

diff --git a/gcc/config.gcc b/gcc/config.gcc
index 6a349965c0a..8641b45662e 100644
--- a/gcc/config.gcc
+++ b/gcc/config.gcc
@@ -3287,6 +3287,13 @@ sparc-*-elf*)
 sparc-*-rtems*)
 	tm_file="${tm_file} dbxelf.h elfos.h sparc/sysv4.h sparc/sp-elf.h sparc/rtemself.h rtems.h newlib-stdint.h"
 	tmake_file="${tmake_file} sparc/t-sparc sparc/t-rtems"
+	case ${target} in
+	    sparc-gaisler*)
+		tm_file="${tm_file} sparc/rcc.h"
+		;;
+	    *)
+		;;
+	esac
 	;;
 sparc-*-linux*)
 	tm_file="${tm_file} dbxelf.h elfos.h sparc/sysv4.h gnu-user.h linux.h glibc-stdint.h sparc/tso.h"
diff --git a/gcc/config/sparc/rcc.h b/gcc/config/sparc/rcc.h
new file mode 100644
index 00000000000..5b6b3a677e9
--- /dev/null
+++ b/gcc/config/sparc/rcc.h
@@ -0,0 +1,37 @@
+/*
+
+This file is part of GCC.
+
+GCC is free software; you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation; either version 3, or (at your option)
+any later version.
+
+GCC is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with GCC; see the file COPYING3.  If not see
+<http://www.gnu.org/licenses/>.  */
+
+#undef STARTFILE_SPEC
+#define STARTFILE_SPEC "%{qrtems|qbsp=*:crti.o%s crtbegin.o%s; :crt0.o%s}"
+
+#undef ENDFILE_SPEC
+#define ENDFILE_SPEC   "%{qrtems|qbsp=*:crtend.o%s crtn.o%s}"
+
+#undef LINK_SPEC
+#define LINK_SPEC      "%{qrtems|qbsp=*:-dc -dp -N}"
+
+#undef DRIVER_SELF_SPECS
+#define DRIVER_SELF_SPECS \
+"%{qrtems:%{!B:%{!qbsp=*:%especify BSP with -qbsp or BSP folder with -B}}} \
+ %{qbsp=*:-B %R/%*/lib}"
+
+#undef LIB_SPEC
+#define LIB_SPEC "%{!qrtems:%{!qbsp=*: " STD_LIB_SPEC "}}" \
+"%{!nostdlib: %{qrtems|qbsp=*: --start-group \
+ -lrtemsbsp -lrtemscpu \
+ -latomic -lc -lgcc --end-group %{!qnolinkcmds: -T linkcmds%s}}}"
diff --git a/gcc/config/sparc/t-rtems b/gcc/config/sparc/t-rtems
index 7d11ab8bfc7..42c53419ff8 100644
--- a/gcc/config/sparc/t-rtems
+++ b/gcc/config/sparc/t-rtems
@@ -18,9 +18,10 @@
 #
 
 MULTILIB_OPTIONS = msoft-float mcpu=v8/mcpu=leon3/mcpu=leon3v7/mcpu=leon \
-		   mfix-ut699/mfix-at697f/mfix-gr712rc
-MULTILIB_DIRNAMES = soft v8 leon3 leon3v7 leon ut699 at697f gr712rc
-MULTILIB_MATCHES = msoft-float=mno-fpu mfix-gr712rc=mfix-ut700
+		   mfix-ut699/mfix-at697f/mfix-gr712rc/mfix-ut700 mflat
+MULTILIB_DIRNAMES = soft v8 leon3 leon3v7 leon mfix-ut699 mfix-at697f \
+		   mfix-gr712rc mfix-ut700 flat
+MULTILIB_MATCHES = msoft-float=mno-fpu
 
 MULTILIB_REQUIRED =
 MULTILIB_REQUIRED += msoft-float
@@ -29,7 +30,8 @@ MULTILIB_REQUIRED += mcpu=leon3
 MULTILIB_REQUIRED += mcpu=leon3v7
 MULTILIB_REQUIRED += mcpu=leon
 MULTILIB_REQUIRED += mcpu=leon3/mfix-gr712rc
-MULTILIB_REQUIRED += mcpu=leon3v7/mfix-gr712rc
+MULTILIB_REQUIRED += mcpu=leon3/mfix-ut700
+MULTILIB_REQUIRED += mcpu=leon3/mflat
 MULTILIB_REQUIRED += mcpu=leon/mfix-ut699
 MULTILIB_REQUIRED += mcpu=leon/mfix-at697f
 MULTILIB_REQUIRED += msoft-float/mcpu=v8
@@ -37,6 +39,7 @@ MULTILIB_REQUIRED += msoft-float/mcpu=leon3
 MULTILIB_REQUIRED += msoft-float/mcpu=leon3v7
 MULTILIB_REQUIRED += msoft-float/mcpu=leon
 MULTILIB_REQUIRED += msoft-float/mcpu=leon3/mfix-gr712rc
-MULTILIB_REQUIRED += msoft-float/mcpu=leon3v7/mfix-gr712rc
+MULTILIB_REQUIRED += msoft-float/mcpu=leon3/mfix-ut700
+MULTILIB_REQUIRED += msoft-float/mcpu=leon3/mflat
 MULTILIB_REQUIRED += msoft-float/mcpu=leon/mfix-ut699
 MULTILIB_REQUIRED += msoft-float/mcpu=leon/mfix-at697f
-- 
2.34.1

